### PR TITLE
add test for empty cluster in nsx_usage.py

### PR DESCRIPTION
**Pull request by gavinbarnard**
add test for empty cluster in nsx_usage.py

### CODE REVIEW

**Comment by yfauser** [comment]
Hi Gavin,

two problems I see, one is minor, which is that your indents use tabs instead of spaces. Can you please make sure to only use spaces as indentation? E.g. PyCharm and Sublime can do this automatically for you. I can correct this later also if you want.
The second one is your change in libutils.py. I'll comment it separately. This changes causes a KeyError when creating logical switches. Can you elaborate what this change is supposed to do?

Thanks!
- Yves

File: pynsxv/library/libutils.py [hunk 1]
```diff
@@ -95,12 +95,13 @@ def get_scope(client_session, transport_zone_name):
     try:
         vdn_scopes = client_session.read('vdnScopes', 'read')['body']
         vdn_scope_list = client_session.normalize_list_return(vdn_scopes['vdnScopes'])
-        vdn_scope = [scope['vdnScope'] for scope in vdn_scope_list
-                     if scope['vdnScope']['name'] == transport_zone_name][0]
+	vdn_scope = [scope['vdnScope'] for scope in vdn_scope_list
+			if scope['vdnScope'][0]['name'] == transport_zone_name][0]
```
Comment (yfauser) [inline pynsxv/library/libutils.py]: scope['vdnScope'] is a dict, so scope['vdnScope'][0] will cause a KeyError. This will lead to the function returning None, None all the time, as the try/except block triggers on KeyErrors. With this change the logical switch creation will never work. What was your intend with this change?

Comment (gavinbarnard) [inline pynsxv/library/libutils.py]: I was err ing out with indice ref must be int not str. On lswitch create


Sent from Samsung Mobile


-------- Original message --------
From: Yves Fauser
Date:08-11-2016 4:25 AM (GMT-05:00)
To: vmware/pynsxv
Cc: Gavin Barnard ,Author
Subject: Re: [vmware/pynsxv] add test for empty cluster in nsx_usage.py (#51)


In pynsxv/library/libutils.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_vmware_pynsxv_pull_51-23discussion-5Fr74383810&d=CwMFaQ&c=Sqcl0Ez6M0X8aeM67LKIiDJAXVeAw-YihVMNtXt-uEs&r=6h6WRoU9RQXcy0gmjS8cmWw41bWf73j9c4rn9HlVRDU&m=KUNa2-tsiiLySx3vy4FMOhLm8_JzEsFAdPTmPRWvq_8&s=iRF8k7kmj0Ki6Y8LycTjajfdoTIpcNhlaSm1Ca2vRW8&e=>:

> @@ -95,12 +95,13 @@ def get_scope(client_session, transport_zone_name):
>      try:
>          vdn_scopes = client_session.read('vdnScopes', 'read')['body']
>          vdn_scope_list = client_session.normalize_list_return(vdn_scopes['vdnScopes'])
> -        vdn_scope = [scope['vdnScope'] for scope in vdn_scope_list
> -                     if scope['vdnScope']['name'] == transport_zone_name][0]
> +     vdn_scope = [scope['vdnScope'] for scope in vdn_scope_list
> +                     if scope['vdnScope'][0]['name'] == transport_zone_name][0]


scope['vdnScope'] is a dict, so scope['vdnScope'][0] will cause a KeyError. This will lead to the function returning None, None all the time, as the try/except block triggers on KeyErrors. With this change the logical switch creation will never work. What was your intend with this change?

-

Comment (gavinbarnard) [inline pynsxv/library/libutils.py]: Here's what happens without my change:

Traceback (most recent call last):
  File "cli.py", line 66, in <module>
    main()
  File "cli.py", line 62, in main
    args.func(args)
  File "/home/cc-admin/grbpynsxv/py2/pynsxv/pynsxv/library/nsx_logical_switch.py", line 216, in _lswitch_main
    logical_switch_name=args.name, verbose=args.verbose)
  File "/home/cc-admin/grbpynsxv/py2/pynsxv/pynsxv/library/nsx_logical_switch.py", line 74, in _logical_switch_create
    logical_switch_name)
  File "/home/cc-admin/grbpynsxv/py2/pynsxv/pynsxv/library/nsx_logical_switch.py", line 48, in logical_switch_create
    vdn_scope_id, vdn_scope = get_scope(client_session, transport_zone)
  File "/home/cc-admin/grbpynsxv/py2/pynsxv/pynsxv/library/libutils.py", line 99, in get_scope
    if scope['vdnScope']['name'] == transport_zone_name][0]
TypeError: list indices must be integers, not str


you can repro with "pynsxv lswitch create -n something"

**Comment by yfauser** [comment]
I see ... You have multiple transport-zone, right?

I think I know where we messed this up :-(. Can you please try with the following:

```python
def get_scope(client_session, transport_zone_name):
    """
    :param client_session: An instance of an NsxClient Session
    :param transport_zone_name: The Name of the Scope (Transport Zone)
    :return: A tuple, with the first item being the scope id as string of the first Scope found with the right name
             and the second item being a dictionary of the scope parameters as return by the NSX API
    """
    try:
        vdn_scopes = client_session.read('vdnScopes', 'read')['body']
        vdn_scope_list = client_session.normalize_list_return(vdn_scopes['vdnScopes']['vdnScope'])
        vdn_scope = [scope for scope in vdn_scope_list if scope['name'] == transport_zone_name][0]
    except KeyError:
        return None, None

    return vdn_scope['objectId'], vdn_scope
```

**Comment by gavinbarnard** [comment]
I'll test out that change later, current engaged in another meeting.

**Comment by yfauser** [comment]
Thanks Gavin!
I merged this into devel. It will be in Master soon, either in the next 0.4.x release or 0.5.
